Feature/114

jaimemosber opened the pull request:
Importación de censo a partir de archivo JSON.

jaimemosber committed:
FEAT: Added import census from JSON
Added feature that allows to import a census from a JSON file

jaimemosber committed:
FIX: Fixed error management
Error messages now appears in the same page when you try to import a census

JaimeBorregoConde commented on decide/census/templates/csv.html:
```diff
@@ -0,0 +1,32 @@
+<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous"> 
+{% block content %}
+<div class="shadow-lg continer">
+  <form method="post" enctype="multipart/form-data">
+    {% csrf_token %}
+ 
+    <div class="row">
+        <div class="col-md-6 col-xs-12">
+          <div class="x_panel">
+            <div class="x_title">
+              <h2>Importación Fichero CSV</h2>
+            
+              <div class="clearfix"></div>
+            </div>
+            <div class="x_content">
+ 
+              <div class="col-md-8 col-sm-12 col-xs-12 form-group">
+                 
+                </label>               
+              </div>
+               
+                    <input type="file" name="myfile" class="form-control">
+                             <button type="submit" class="btn btn-success" >Subir</button>                            
+                            <br/>
+                            </div>
+                        </div>
+                    </div>
+                </div>
+  </form>
+  </div>
+    
+{% endblock %}
```
Elimina esto de aquí que no es de esta rama y puede dar conflictos cuando se haga el merge al módulo.

jaimemosber replied:
He hecho un nuevo commit y lo he eliminado.

JaimeBorregoConde replied:
Correcto, ha desaparecido

JaimeBorregoConde commented on decide/census/templates/json.html:
```diff
@@ -0,0 +1,40 @@
+<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous"> 
+{% block content %}
+
+
+{% if messages %}
+    {% for message in messages %}
+    <div {% if message.tags == 'error' %} class="alert alert-danger" {% elif message.tags == 'success' %} class="alert alert-success"{% endif %}  role="alert">
+      {{ message }}</div>
+    {% endfor %}
+{% endif %}
+<div class="shadow-lg continer">
+  <form method="post" enctype="multipart/form-data">
+    {% csrf_token %}
+ 
+    <div class="row">
+        <div class="col-md-6 col-xs-12">
+          <div class="x_panel">
+            <div class="x_title">
+              <h2>Importación Fichero JSON</h2>
+            
+              <div class="clearfix"></div>
+            </div>
+            <div class="x_content">
+ 
+              <div class="col-md-8 col-sm-12 col-xs-12 form-group">
+                 
+                </label>               
+              </div>
+               
+                    <input type="file" name="myfile" class="form-control">
+                             <button type="submit" class="btn btn-success" >Subir</button>                            
+                            <br/>
+                            </div>
+                        </div>
+                    </div>
+                </div>
+  </form>
+  </div>
+    
+{% endblock %}
```
He observado que tienes una tarea para diseño de la importación de json, por lo que asumo que el archivo json.html no es el definitivo. Si no es el caso, tienes que añadir las etiquetas de traducción (<!--TRADUCCION-->)

jaimemosber replied:
No es el json.html definitivo

JaimeBorregoConde replied:
ok, entonces todo en orden.

JaimeBorregoConde commented on decide/census/views.py:
```diff
@@ -34,7 +52,85 @@ def list(self, request, *args, **kwargs):
+class CensusImport(generics.ListCreateAPIView):
+    permission_classes = (UserIsStaff,)
+
+    @transaction.atomic
+    @api_view(['GET','POST'])
+    def import_csv(request):
+        cont=2
+        try: 
+            if request.method == 'POST':
+                census_from_csv=[]
+            
+                myfile = request.FILES['myfile'] 
+                df=pd.read_csv(myfile)
+
+                for d in df.values:
+                    try:
+                        group = None
+                        if not math.isnan(d[2]):
+                            group = CensusGroup.objects.get(id=d[2])
+
+                        census = Census(voting_id=d[0], voter_id=d[1],group=group)
+                        census_from_csv.append(census)
+                        cont+=1
+                    except CensusGroup.DoesNotExist:
+                        return Response('The input Census Group does not exist, in row {}'.format(cont-1), status=ST_400)
+                cont=0
+                for c in census_from_csv:
+                    try:
+                        cont+=1
+                        c.save()
+                    except IntegrityError:
+                        return Response('Error trying to import CSV, in row {}. A census cannot be repeated.'.format(cont), status=ST_409)
+                return Response('Census created', status=ST_201)
+        except:
+            return Response('Error in CSV data. There are wrong data in row {}'.format(cont+1), status=ST_409)
+        return render(request,"csv.html")
+
```
elimina esto de aquí para tener las ramas separadas y que no de conflictos al hacer el merge

jaimemosber replied:
Lo he eliminado y he hecho un nuevo commit.

JaimeBorregoConde replied:
Perfecto, probado y funciona a la perfección.

JaimeBorregoConde commented on decide/census/views.py:
```diff
@@ -34,7 +52,85 @@ def list(self, request, *args, **kwargs):
+            return Response('Error in CSV data. There are wrong data in row {}'.format(cont+1), status=ST_409)
+        return render(request,"csv.html")
+
+    @transaction.atomic
+    @api_view(['GET','POST'])
+    def import_json(request):
+        cont=2
+        try: 
+            if request.method == 'POST':
+                census_from_json=[]
+            
+                myfile = request.FILES['myfile'] 
+                df=pd.read_json(myfile)
+
+                for d in df.values:
+                    try:
+                        group= None
+                        if d[2]:
+                            if d[2] == "":
+                                group = None
+                            else: 
+                                group = CensusGroup.objects.get(id=d[2])
+
+                        census = Census(voting_id=d[0], voter_id=d[1],group=group)
+                        census_from_json.append(census)
+                    except CensusGroup.DoesNotExist:
+                        messages.error(request,'The input Census Group does not exist')
+                        return render(request,"json.html")
+                for c in census_from_json:
+                    try:
+                        c.save()
+                    except IntegrityError:
+                        messages.error(request, 'Error trying to import JSON. A census cannot be repeated.')
+                        return render(request,"json.html")
+                messages.success(request, 'Census created')
+        except:
+            messages.error(request, 'Error in JSON data.') 
+            return render(request,"json.html")
+        return render(request,"json.html")
+
```
Por otra parte, en el views.py de census, la clase donde está definida la función import_json en incorrecta, ya que es una vista de API. Deberías o cambiar la clase o sacar la función y eliminar la línea de @api_view

JaimeBorregoConde replied:
Perfecto, probado y funciona a la perfección.

jaimemosber committed:
FIX: Import census from JSON function fixed and some files and functions deleted
Import census from JSON function fixed an error because is not an API. I have also deleted csv.html and import census from csv function because it's from another feature

JaimeBorregoConde committed:
Fix: Problems in merge
Merging feature/114 with module/censo

JaimeBorregoConde reviewed (approved):
Todo arreglado y funcionando
